diff --git a/gdk/macos/gdkmacossurface-private.h b/gdk/macos/gdkmacossurface-private.h
index 4450863388..ca72ef54f5 100644
--- a/gdk/macos/gdkmacossurface-private.h
+++ b/gdk/macos/gdkmacossurface-private.h
@@ -65,6 +65,7 @@ struct _GdkMacosSurface
 
   gint64 pending_frame_counter;
 
+  guint initial_update_freeze : 1;
   guint did_initial_present : 1;
   guint geometry_dirty : 1;
   guint next_frame_set : 1;
diff --git a/gdk/macos/gdkmacossurface.c b/gdk/macos/gdkmacossurface.c
index 388c8f6799..0eb5069549 100644
--- a/gdk/macos/gdkmacossurface.c
+++ b/gdk/macos/gdkmacossurface.c
@@ -484,6 +484,7 @@ gdk_macos_surface_constructed (GObject *object)
                                G_CONNECT_SWAPPED);
     }
 
+  self->initial_update_freeze = TRUE;
   gdk_surface_freeze_updates (GDK_SURFACE (self));
   _gdk_macos_surface_monitor_changed (self);
 
@@ -767,7 +768,11 @@ _gdk_macos_surface_show (GdkMacosSurface *self)
       gdk_surface_set_is_mapped (GDK_SURFACE (self), TRUE);
       gdk_surface_request_layout (GDK_SURFACE (self));
       gdk_surface_invalidate_rect (GDK_SURFACE (self), NULL);
-      gdk_surface_thaw_updates (GDK_SURFACE (self));
+      if (self->initial_update_freeze)
+        {
+          self->initial_update_freeze = FALSE;
+          gdk_surface_thaw_updates (GDK_SURFACE (self));
+        }
     }
 }
 
diff --git a/testsuite/gtk/regression-tests.c b/testsuite/gtk/regression-tests.c
index 2187c46148..2200f1af10 100644
--- a/testsuite/gtk/regression-tests.c
+++ b/testsuite/gtk/regression-tests.c
@@ -48,12 +48,36 @@ test_94f00eb04dd1433cf1cc9a3341f485124e38abd1 (void)
   gtk_window_destroy (GTK_WINDOW (window));
 }
 
+static void
+test_toplevel_remap (void)
+{
+  GtkWidget *window;
+  guint cycle;
+
+  window = gtk_window_new ();
+
+  gtk_widget_set_visible (window, TRUE);
+  g_assert_true (gtk_widget_get_mapped (window));
+
+  for (cycle = 0; cycle < 3; cycle++)
+    {
+      gtk_widget_set_visible (window, FALSE);
+      g_assert_false (gtk_widget_get_mapped (window));
+
+      gtk_widget_set_visible (window, TRUE);
+      g_assert_true (gtk_widget_get_mapped (window));
+    }
+
+  gtk_window_destroy (GTK_WINDOW (window));
+}
+
 int
 main (int argc, char *argv[])
 {
   gtk_test_init (&argc, &argv, NULL);
 
   g_test_add_func ("/regression/94f00eb04dd1433cf1cc9a3341f485124e38abd1", test_94f00eb04dd1433cf1cc9a3341f485124e38abd1);
   g_test_add_func ("/regression/9d6da33ff5c5e41e3521e1afd63d2d67bc915753", test_9d6da33ff5c5e41e3521e1afd63d2d67bc915753);
+  g_test_add_func ("/regression/toplevel-remap", test_toplevel_remap);
 
   return g_test_run ();
 }
